sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowGroupLimitExec.scala

```diff
@@ -0,0 +1,310 @@
+    case _: RowNumber if partitionSpec.isEmpty =>
+      child.execute().mapPartitionsInternal(new SimpleLimitIterator(output, _, limit))
+    case _: RowNumber =>
+      child.execute().mapPartitionsInternal(
+        SimpleGroupLimitIterator(partitionSpec, output, _, limit))
+    case _: Rank if partitionSpec.isEmpty =>
+      child.execute().mapPartitionsInternal(new RankLimitIterator(output, _, orderSpec, limit))
+    case _: Rank =>
+      child.execute().mapPartitionsInternal(
+        RankGroupLimitIterator(partitionSpec, output, _, orderSpec, limit))
+    case _: DenseRank if partitionSpec.isEmpty =>
+      child.execute().mapPartitionsInternal(new DenseRankLimitIterator(output, _, orderSpec, limit))
+    case _: DenseRank =>
+      child.execute().mapPartitionsInternal(
+        DenseRankGroupLimitIterator(partitionSpec, output, _, orderSpec, limit))
+  }
+
+  override protected def withNewChildInternal(newChild: SparkPlan): WindowGroupLimitExec =
+    copy(child = newChild)
+}
+
+abstract class BaseLimitIterator extends Iterator[InternalRow] {
+
+  def output: Seq[Attribute]
+
+  def input: Iterator[InternalRow]
+
+  def limit: Int
+
+  var rank = 0
+
+  var nextRow: UnsafeRow = null
+
+  // Increase the rank value.
+  def increaseRank(): Unit
+
+  override def hasNext: Boolean = rank < limit && input.hasNext
+
+  override def next(): InternalRow = {
+    nextRow = input.next().asInstanceOf[UnsafeRow]
+    increaseRank()
+    nextRow
+  }
+}
+
+class SimpleLimitIterator(
+    val output: Seq[Attribute],
+    val input: Iterator[InternalRow],
+    val limit: Int) extends BaseLimitIterator {
+
+  override def increaseRank(): Unit = {
+    rank += 1
+  }
+}
+
+trait OrderSpecProvider {
+  def output: Seq[Attribute]
+  def orderSpec: Seq[SortOrder]
+
+  val ordering = GenerateOrdering.generate(orderSpec, output)
+  var currentRankRow: UnsafeRow = null
+}
+
+class RankLimitIterator(
+    val output: Seq[Attribute],
+    val input: Iterator[InternalRow],
+    val orderSpec: Seq[SortOrder],
+    val limit: Int) extends BaseLimitIterator with OrderSpecProvider {
+
+  var count = 0
+
+  override def increaseRank(): Unit = {
+    if (count == 0) {
+      currentRankRow = nextRow.copy()
+    } else {
+      if (ordering.compare(currentRankRow, nextRow) != 0) {
+        rank = count
+        currentRankRow = nextRow.copy()
+      }
+    }
+    count += 1
+  }
+}
+
+class DenseRankLimitIterator(
+    val output: Seq[Attribute],
+    val input: Iterator[InternalRow],
+    val orderSpec: Seq[SortOrder],
+    val limit: Int) extends BaseLimitIterator with OrderSpecProvider {
+
+  override def increaseRank(): Unit = {
+    if (currentRankRow == null) {
+      currentRankRow = nextRow.copy()
+    } else {
+      if (ordering.compare(currentRankRow, nextRow) != 0) {
+        rank += 1
+        currentRankRow = nextRow.copy()
+      }
+    }
+  }
+}
+
+trait WindowIterator extends BaseLimitIterator {
```

Comment:
I think the main job of this iterator is to do grouping, and then we can build a `SimpleLimitIterator` or its friends with each group. Something like
```
class GroupIterator(input, currentRow) extends Iterator[InternalRow] {
  val currentGroup = grouping(currentRow)
  // Manage the stream and the grouping.
  var nextGroup: UnsafeRow = null
  var nextRow: UnsafeRow = null
  protected[this] def fetchNextRow(): Unit = {
    if (input.hasNext) {
      nextRow = input.next().asInstanceOf[UnsafeRow]
      nextGroup = grouping(nextRow)
    } else {
      nextRow = null
      nextGroup = null
    }
  }
  fetchNextRow()

  def hasNext = nextGroup != null && nextGroup == currentGroup

  def next = {
    val res = nextRow
    fetchNextRow()
    res
  }
}


trait WindowIterator extends Iterator[InternalRow] {

  var limitIterator = null
  def createLimitIterator = {
    if (input.hasNext) {
      val groupIterator = new GroupIterator(input, itput.next)
      limitIterator = createSimpleLimitIterator(groupIterator)
    } else {
      limitIterator = null
    }
  }

  createLimitIterator()
  
  def hasNext = limitIterator != null && limitIterator.hasNext
  def next = {
    val res = limitIterator.next()
    if (!limitIterator.hasNext) createLimitIterator
    res
  }
}
```